sdk/python/tests/conftest.py

```diff
@@ -161,86 +165,152 @@ def start_test_local_server(repo_path: str, port: int):
     fs.serve("localhost", port, no_access_log=True)
 
 
-@pytest.fixture(
-    params=FULL_REPO_CONFIGS, scope="session", ids=[str(c) for c in FULL_REPO_CONFIGS]
-)
-def environment(request, worker_id: str):
+@pytest.fixture(scope="session")
+def environment(request, worker_id):
     e = construct_test_environment(
         request.param, worker_id=worker_id, fixture_request=request
     )
+
+    yield e
+
+    e.feature_store.teardown()
+    e.data_source_creator.teardown()
+    if e.online_store_creator:
+        e.online_store_creator.teardown()
+
+
+_config_cache = {}
+
+
+def pytest_generate_tests(metafunc):
```

Comment:
Can you add some docs to how this `pytest_generate_tests` thingy works? It's the first time I've seen it used and seems extremely powerful

Reply:
Added comment.